File: day12/main.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"log"
	"os"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	sum := 0
	sum2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		a, b, ok := bytes.Cut(scanner.Bytes(), []byte{' '})
		if !ok {
			log.Fatalln("Invalid line")
		}
		var nums []int
		for _, i := range bytes.Split(b, []byte{','}) {
			num, err := strconv.Atoi(string(i))
			if err != nil {
				log.Fatalln(err)
			}
			nums = append(nums, num)
		}
		cache := make([]int, len(a)*len(nums))
		sum += getNumArrangements(a, nums, cache, len(a))
		aa := make([]byte, 0, len(a)*5+5)
		bb := make([]int, 0, len(nums)*5)
		for i := 0; i < 5; i++ {
			if len(aa) > 0 {
				aa = append(aa, '?')
			}
			aa = append(aa, a...)
			bb = append(bb, nums...)
		}
		cache = make([]int, len(aa)*len(bb))
		sum2 += getNumArrangements(aa, bb, cache, len(aa))
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", sum)
	fmt.Println("Part 2:", sum2)
}

func getNumArrangements(b []byte, nums []int, cache []int, cacheRowWidth int) int {
	if len(nums) == 0 {
		if restNoGroup(b) {
			return 1
		} else {
			return 0
		}
	}
	if len(b) == 0 {
		return 0
	}

	key := cacheRowWidth*(len(nums)-1) + len(b) - 1
	if n := cache[key]; n > 0 {
		return n - 1
	}

	first := b[0]
	if first == '.' {
		count := getNumArrangements(b[1:], nums, cache, cacheRowWidth)
		cache[key] = count + 1
		return count
	}

	firstNum := nums[0]
	prefixMatches, isEnd := matchPrefix(b, firstNum)
	if isEnd {
		if len(nums) == 1 {
			cache[key] = 2
			return 1
		}
		cache[key] = 1
		return 0
	}

	count := 0
	if first == '?' {
		count += getNumArrangements(b[1:], nums, cache, cacheRowWidth)
	} else {
		if !prefixMatches {
			cache[key] = 1
			return 0
		}
	}
	if prefixMatches {
		count += getNumArrangements(b[firstNum+1:], nums[1:], cache, cacheRowWidth)
	}

	cache[key] = count + 1

	return count
}

func restNoGroup(b []byte) bool {
	for _, i := range b {
		if i == '#' {
			return false
		}
	}
	return true
}

func matchPrefix(b []byte, num int) (bool, bool) {
	if len(b) < num {
		return false, false
	}
	for i := 0; i < num; i++ {
		if b[i] == '.' {
			return false, false
		}
	}
	if len(b) == num {
		return true, true
	}
	if b[num] == '#' {
		return false, false
	}
	return true, false
}

File: day12/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut sum = 0;
    let mut sum2 = 0;
    for line in reader.lines() {
        let line = line?;
        let (a, b) = line.split_once(' ').ok_or("Invalid line")?;
        let nums = b
            .split(',')
            .map(|v| v.parse::<usize>())
            .collect::<Result<Vec<_>, _>>()?;
        sum += get_num_arrangements(
            a.as_bytes(),
            &nums,
            &mut vec![0; a.len() * nums.len()],
            a.len(),
        );
        let ai = a.bytes().collect::<Vec<_>>();
        let mut aa = ai.clone();
        let mut bb = nums.clone();
        for _ in 0..4 {
            aa.push(b'?');
            aa.append(&mut ai.clone());
            bb.append(&mut nums.clone());
        }
        sum2 += get_num_arrangements(&aa[..], &bb, &mut vec![0; aa.len() * bb.len()], aa.len());
    }

    println!("Part 1: {}\nPart 2: {}", sum, sum2);

    Ok(())
}

fn get_num_arrangements(
    b: &[u8],
    nums: &[usize],
    cache: &mut Vec<usize>,
    cache_row_width: usize,
) -> usize {
    if nums.len() == 0 {
        return if rest_no_group(b) { 1 } else { 0 };
    }
    if b.len() == 0 {
        return 0;
    }

    let key = cache_row_width * (nums.len() - 1) + b.len() - 1;
    if cache[key] > 0 {
        return cache[key] - 1;
    }

    let first = b[0];
    if first == b'.' {
        let count = get_num_arrangements(&b[1..], nums, cache, cache_row_width);
        cache[key] = count + 1;
        return count;
    }

    let first_num = nums[0];
    let (prefix_matches, is_end) = match_prefix(b, first_num);
    if is_end {
        return if nums.len() == 1 {
            cache[key] = 2;
            1
        } else {
            cache[key] = 1;
            0
        };
    }

    let mut count = 0;
    if first == b'?' {
        count += get_num_arrangements(&b[1..], nums, cache, cache_row_width);
    } else {
        if !prefix_matches {
            cache[key] = 1;
            return 0;
        }
    }

    if prefix_matches {
        count += get_num_arrangements(&b[first_num + 1..], &nums[1..], cache, cache_row_width);
    }

    cache[key] = count + 1;

    count
}

fn rest_no_group(b: &[u8]) -> bool {
    b.iter().all(|&v| v != b'#')
}

fn match_prefix(b: &[u8], num: usize) -> (bool, bool) {
    if b.len() < num {
        return (false, false);
    }
    for i in 0..num {
        if b[i] == b'.' {
            return (false, false);
        }
    }
    if b.len() == num {
        (true, true)
    } else if b[num] == b'#' {
        (false, false)
    } else {
        (true, false)
    }
}
